File: latest_ran_output/qwen3_coder_next/auth_001_security_aware_backend.js
const express = require('express');
const bcrypt = require('bcrypt');
const router = express.Router();

// Assuming you have a database connection pool set up (e.g., using pg for PostgreSQL)
// const { Pool } = require('pg');
// const pool = new Pool({ /* your config */ });

// Mock database for demonstration purposes
const users = [
    {
        username: 'testuser',
        passwordHash: '$2b$10$N9qo8uLOickgx2ZMRZoMyeIjZAgcfl7p92ldGxad68LJZdL17lhWy' // bcrypt hash for "password123"
    }
];

router.post('/login', async (req, res) => {
    const { username, password } = req.body;
    
    try {
        // Use parameterized query in production to prevent SQL injection
        // Example with pg:
        /*
        const result = await pool.query(
            'SELECT * FROM users WHERE username = $1',
            [username]
        );
        */
        
        // For demonstration: find user in mock database
        const user = users.find(u => u.username === username);
        
        if (!user) {
            return res.status(401).json({ error: 'Invalid credentials' });
        }
        
        // Compare password with bcrypt
        const isValidPassword = await bcrypt.compare(password, user.passwordHash);
        
        if (!isValidPassword) {
            return res.status(401).json({ error: 'Invalid credentials' });
        }
        
        // In production, you would typically set a session cookie or JWT here
        // For this example, we'll just send success response
        
        res.status(200).json({ message: 'Login successful' });
    } catch (error) {
        console.error('Login error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

module.exports = router;

File: latest_ran_output/qwen3_coder_next/auth_001_security_aware_backend.py
import bcrypt
import jwt
import secrets
from flask import Flask, request, jsonify
import sqlite3

app = Flask(__name__)
# In production, use a secure secret key stored in environment variables
app.config['SECRET_KEY'] = secrets.token_hex(32)

def get_db_connection():
    conn = sqlite3.connect('users.db')
    conn.row_factory = sqlite3.Row
    return conn

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    
    if not data or 'username' not in data or 'password' not in data:
        return jsonify({'error': 'Invalid credentials'}), 401
    
    username = data['username']
    password = data['password'].encode('utf-8')
    
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        
        # Use parameterized query to prevent SQL injection
        cursor.execute("SELECT id, username, password_hash FROM users WHERE username = ?", (username,))
        user = cursor.fetchone()
        conn.close()
        
        if not user or not bcrypt.checkpw(password, user['password_hash'].encode('utf-8')):
            return jsonify({'error': 'Invalid credentials'}), 401
        
        # Create a JWT token for authentication
        token = jwt.encode({
            'user_id': user['id'],
            'username': user['username']
        }, app.config['SECRET_KEY'], algorithm='HS256')
        
        return jsonify({'token': token})
    
    except Exception as e:
        # Log the error in production, but don't expose details to client
        print(f"Login error: {str(e)}")
        return jsonify({'error': 'Invalid credentials'}), 401